### PR TITLE
perf: typed RDF literal storage — indexed WHERE, boot migrations, envelope-free writes

### PR DESCRIPTION
**Pull request by HexaField**
## What this does

Properties with `resolveLanguage: "literal"` (the default) now store their values as XSD-typed RDF literals in Oxigraph instead of NamedNode IRIs. This means WHERE clauses hit the POS index directly instead of parsing strings through FILTER, and writes skip the `expression_create` round-trip entirely.

### Storage

- `literal:string:X` → `"X"^^xsd:string`
- `literal:number:N` → `"N"^^xsd:integer` or `xsd:decimal`
- `literal:boolean:B` → `"B"^^xsd:boolean`
- `literal:json:J` → `"J"^^ad4m://json`
- Everything else (DIDs, URIs, language expressions) stays as NamedNode IRIs

### Queries

- WHERE conditions emit typed-literal triple patterns (`"val"^^<xsd:string>`) instead of `FILTER(STR(fn/parse_literal(?x)) = "val")`
- ORDER BY uses `STR(?raw)` directly instead of `STR(fn/parse_literal(?raw))`
- `fn/parse_literal` is still registered as a SPARQL function for custom queries

### Writes

- `setProperty` for literal-resolved properties calls `valueToLiteralIri()` and writes directly — no `expression_create`, no signed envelope
- Auto-generated base expressions use `ad4m://obj/<id>` instead of `literal:string:<id>` so instance IDs stay as IRIs and don't get typed-literal treatment

### Boot migrations

Two migrations run on perspective load (skip + retry on failure):
1. Signed envelopes → plain `literal:*` targets
2. IRI-shaped `literal:*` targets → XSD-typed literals

### Other

- `projection.rs` simplified — `fn/parse_literal` handles all types uniformly, removed `typed_number_literal` branching and per-type XSD constants
- JSDoc on `@Property`/`@Flag` decorators updated to describe the storage behaviour

## Summary by CodeRabbit

* **New Features**
  * Added deterministic support for boolean, string, number, and JSON literals.
  * Model instances now receive unique identifiers, even when their properties match.
  * Added AND, OR, and NOT query filters, plus expanded sorting for relation properties and projection counts.
  * Improved filtering and pagination for typed literal values.

* **Bug Fixes**
  * Improved literal parsing, encoding, hydration, and legacy-value migration.
  * Strengthened SPARQL escaping to prevent unsafe mention matching.

* **Documentation**
  * Clarified literal resolution, model identity, and uniqueness behavior.

### CODE REVIEW

File: rust-executor/src/mcp/tools/subscriptions.rs [hunk 1]
```diff
@@ -213,19 +213,11 @@ impl Ad4mMcpHandler {
...
         let mention_conditions: Vec<String> = all_terms
             .iter()
-            .map(|t| {
-                format!(
-                    "CONTAINS(LCASE(STR(<ad4m://fn/parse_literal>(?target))), \"{}\")",
-                    t
-                )
-            })
+            .map(|t| format!("CONTAINS(LCASE(STR(?target)), \"{}\")", t))
             .collect();
```
[the remark this answers is not in the document]
Comment (HexaField) [reply rust-executor/src/mcp/tools/subscriptions.rs:243]: Fixed in dafc73461 — mention terms (profile names / `name_override`) are now escaped before interpolation into the SPARQL string literal (backslash/quote/control chars), matching the escaping pattern already used elsewhere in the codebase. Added two regression tests confirming an escaped payload can't reintroduce an unescaped quote/break out of the literal.

File: rust-executor/src/perspectives/model_query/integration_tests.rs [hunk 2]
```diff
@@ -4887,3 +5086,509 @@ async fn test_sort_by_relation_property_with_missing_relation() {
...
+            .unwrap();
+    }
+
+    let indexed = format!("SELECT ?source WHERE {{ ?source <{pred}> \"{target_value}\" . }}");
+    let filtered = format!(
+        "SELECT ?source WHERE {{ \
+            ?source <{pred}> ?t . \
+            FILTER(STR(?t) = \"{target_value}\") \
+        }}"
+    );
```
[the remark this answers is not in the document]
Comment (HexaField) [reply rust-executor/src/perspectives/model_query/integration_tests.rs]: Fixed in 74293c25d — renamed to `bench_indexed_literal_vs_str_filter` (and updated the `cargo test` invocation hint) and corrected the header comment to describe the actual `FILTER(STR(?t) = ...)` comparison rather than `fn/parse_literal`.

File: tests/js/tests/prolog-and-literals.test.ts [hunk 3]
```diff
@@ -2206,10 +2215,20 @@ describe("Prolog + Literals", () => {
...
+                        // change (#848), which adds first-fetch latency on a
+                        // freshly registered SDNA class. Steady-state
+                        // subscription latency is still logged via
+                        // `subscriptionLatency`, so a real regression would
+                        // surface as a slow log line rather than be hidden by
+                        // the bumped ceiling.
                         while (!subscriptionCallback.called) {
                             await sleep(10);
-                            if (Date.now() - saveTime > 5000) throw new Error("Timeout waiting for subscription update");
+                            if (Date.now() - saveTime > 60000) throw new Error("Timeout waiting for subscription update");
```
[the remark this answers is not in the document]
Comment (HexaField) [reply tests/js/tests/prolog-and-literals.test.ts:2231]: Fixed in 74293c25d — comment now says 5s, matching the actual prior ceiling.

File: rust-executor/src/perspectives/sparql_store.rs [hunk 4]
```diff
@@ -17,6 +17,118 @@ const ONT_PROOF_VALID: &str = "ad4m://ontology/proofValid";
...
+            // Re-serialise so the stored lexical form is canonical JSON; if
+            // parsing fails, keep the decoded text under xsd:string so the
+            // value at least round-trips even though it's not JSON.
+            if let Ok(json_val) = serde_json::from_str::<Value>(&decoded) {
+                let canonical = serde_json::to_string(&json_val).unwrap_or(decoded);
+                return Literal::new_typed_literal(canonical, NamedNode::new_unchecked(ONT_JSON))
+                    .into();
+            }
+            return Literal::new_simple_literal(decoded).into();
+        }
```
[the remark this answers is not in the document]
Comment (HexaField) [reply rust-executor/src/perspectives/sparql_store.rs:94]: Fixed in 2242674d3 — `make_reifier_iri` now hashes the normalized storage-term representation (`storage_term_to_target_string(&target_to_storage_term(...))`) instead of the raw wire string, so a hydrate-then-remove round trip recomputes the same reifier IRI. Added a regression test covering exactly that round trip (insert with whitespace, hydrate via query, remove with the hydrated string).

One caveat worth flagging explicitly: this crate builds `serde_json` with the `preserve_order` feature (pulled in transitively via `schemars`), so "canonical" here normalizes whitespace/formatting, not object key order — two *independently hand-written* JSON strings with the same keys in different orders would still diverge. That's narrower than a literal reading of "canonicalized" might suggest, but it matches what your own repro script actually demonstrated (whitespace, via a hydrate round-trip), and it's the realistic failure mode (read-then-remove) rather than the more contrived one (two independently-authored orderings). Happy to open a follow-up for true key-order-independent canonicalization if that's wanted — it'd change the stored lexical form for new JSON writes, so I didn't want to bundle it into this fix unprompted.

File: rust-executor/src/perspectives/sparql_store.rs [hunk 5]
```diff
@@ -817,9 +1004,27 @@ impl SparqlStore {
...
                             let val = match term {
                                 Term::NamedNode(n) => Value::String(n.as_str().to_string()),
-                                Term::Literal(l) => Value::String(l.value().to_string()),
+                                Term::Literal(l) => {
+                                    if is_target_var {
+                                        Value::String(storage_term_to_target_string(term))
+                                    } else {
+                                        Value::String(l.value().to_string())
+                                    }
+                                }
```
[the remark this answers is not in the document]
Comment (HexaField) [reply rust-executor/src/perspectives/sparql_store.rs:1072]: Addressed, but scoped narrower than renaming the `?target`/`?t` convention everywhere — that convention turned out to be pervasive (perspective_instance.rs getters, projection.rs, query.rs, shape.rs, relations.rs, getters.rs, plus test fixtures hardcoding exact query strings), so a blanket rename carried real regression risk for a Minor-severity, narrow edge case.

Instead I traced every caller of `SparqlStore::query()` and found exactly one arbitrary/caller-supplied-query entrypoint: `perspective_instance::sparql_query()`, reached via the `perspective.querySparql` RPC. Split into `query()` (unchanged — all ~15 internal callers keep wire-format hydration, zero risk to them) and a new `query_arbitrary()` (no hydration) used only by that one entrypoint. Left the notification-trigger call site (`perspective_instance.rs` around line 3269) on `query()`, since notification triggers are domain-specific queries about links/targets and more likely to intentionally rely on the `?target` convention than to hit this collision by accident.

Regression test added confirming `query()` and `query_arbitrary()` now diverge correctly for a coincidentally-named `?t`.

File: core/src/model/types.ts [hunk 6]
```diff
@@ -380,8 +380,13 @@ export interface PropertyMetadata {
...
   /** Initial value if specified */
   initial?: string;
-  /** Language for resolution (e.g., "literal") */
+  /** Language used to resolve the value (e.g. "literal" or a custom language
+   *  address). Custom languages route values through expression_create. */
   resolveLanguage?: string;
+  /** When true (default) and resolveLanguage is "literal", values are stored
+   *  as deterministic literal: IRIs. When false, the literal value goes
+   *  through expression_create on the literal language. */
+  resolveLiteral?: boolean;
```
[the remark this answers is not in the document]
Comment (HexaField) [reply core/src/model/types.ts]: This field (`resolveLiteral`) no longer exists in this file — the PR later converged on a single `resolveLanguage` field (`"literal"`/unset → deterministic `literal:` storage; anything else → signed-envelope expression), which is what's actually in the diff now. The current `resolveLanguage` doc comment at this location is accurate as written. Resolving as moot rather than editing — there's nothing left to fix.

File: rust-executor/src/perspectives/model_query/types.rs [hunk 7]
```diff
@@ -359,7 +359,17 @@ pub struct ShapeProperty {
...
+    /// mechanism). `Some("literal")` is the built-in literal language;
+    /// any other address routes values through `expression_create` on that
+    /// language. `None` means values are stored raw.
     pub(crate) resolve_language: Option<String>,
+    /// Optimization flag for the literal language: `Some(true)` (the default)
+    /// stores values as deterministic `literal:*` IRIs for POS-index lookups;
+    /// `Some(false)` routes the literal value through `expression_create`,
+    /// producing a signed-envelope URI. Only meaningful when
+    /// `resolve_language` is `"literal"` / unset.
+    pub(crate) resolve_literal: Option<bool>,
```
[the remark this answers is not in the document]
Comment (HexaField) [reply rust-executor/src/perspectives/model_query/types.rs]: Same as the `core/src/model/types.ts` thread — `resolve_literal` no longer exists in this file, superseded by `resolve_language: Option<String>`, which is already documented correctly at its current location. Resolving as moot.

**Comment by HexaField** [comment]
Closing in favour of #874 + #876.

#874 carries the same typed-RDF-literal storage engine with the more flexible API shape (`resolveLiteral` boolean for per-property choice between deterministic literals and signed envelopes). #876 cherry-picks all the orthogonal follow-up fixes from this branch (SPARQL injection, reifier IRI stability, query/query_arbitrary split, Ops handler, encoding alignment, uniqueness tests).

The stacked PRs (#812 named graphs, #873 graph-expression duality) have been restacked onto #876 → #812 → #873. Build-verified, zero conflicts on the named-graph/graph-expression work since it has no dependency on the storage-mode API shape.

⬡
